### PR TITLE
Issue 91 - CSS Optimization

### PR DESCRIPTION
**Pull request by dlarson04**
Signed-off-by: Doug Larson <larsond@us.ibm.com>

- Use a different httpClient object for agent with a longer timeout for object download requests
- Implement a semaphore on handleGetData instead of a lock to allow some concurrent model downloads but not unlimited
- Change so that if an error occurs in the model download, don't allow the CSS to set the notification record

### CODE REVIEW

File: common/config.go [hunk 1]
```diff
@@ -204,6 +204,14 @@ type Config struct {
 	// default is 120s
 	HTTPESSClientTimeout int `env:"HTTPESSClientTimeout"`
 
+	// HTTPESSObjClientTimeout is to specify the http client timeout for downloading models (or objects) in seconds for ESS
+	// default is 600s
+	HTTPESSObjClientTimeout int `env:"HTTPObjESSClientTimeout"`
+
+	// HTTPESSObjDownloadConcurrencyMultiplier specifies a number to multiple the number of threads by to set allowed concurrent downloads per CSS
+	// default is 1
+	HTTPESSObjDownloadConcurrencyMultiplier int `env:"HTTPESSObjDownloadConcurrencyMultiplier"`
```
Comment (dabooz) [inline common/config.go:213]: nit, this should probably be named for the CSS, instead of ESS.

File: core/base/base.go [hunk 2]
```diff
@@ -145,14 +145,12 @@ func Start(swaggerFile string, registerHandlers bool) common.SyncServiceError {
 
 	if common.Configuration.NodeType == common.ESS {
 		common.Registered = false
-		if common.Configuration.CommunicationProtocol == common.HTTPProtocol {
-			go communication.Register()
```
Comment (dabooz) [inline core/base/base.go:149]: Why did you remove this?

File: core/communications/httpCommunication.go [hunk 3]
```diff
@@ -1158,36 +1187,27 @@ func (communication *HTTP) handleGetData(orgID string, objectType string, object
...
-				}
-			} else {
-				if trace.IsLogging(logger.DEBUG) {
-					trace.Debug("Handling object get data, return without update notification status to data for %s %s %s %s %s, set updateNotificationRecord to %t \n", orgID, objectType, objectID, destType, destID, updateNotificationRecord)
-				}
+				handleGetData_download_complete_time := time.Now().Unix()
+				trace.Debug("handleGetData has taken %d seconds to downloads %s %s %s %s", (handleGetData_download_complete_time - handleGetData_enter_time), orgID, objectType, objectID, destID)
 			}
+
+			/**
```
Comment (dabooz) [inline core/communications/httpCommunication.go:1204]: This code was integral to the last set of performance changes, to ensure that all the models were correctly delivered. Removal of this doesnt open up the possibility of leaving the node in an error state when the model was actually delivered correctly?

**Review by dabooz — CHANGES_REQUESTED** [review]

File: core/communications/httpCommunication.go (hunk 3 above)
Comment (dlarson04) [reply core/communications/httpCommunication.go:1204]: In my testing that ran, I did not encounter any problems when I removed this code and going up to the 2500 agents again. And I did experience intermittent failures when I had it in.

Comment (dabooz) [reply core/communications/httpCommunication.go:1204]: ok

File: common/config.go (hunk 1 above)
Comment (dlarson04) [reply common/config.go]: Updated code with the change

File: core/base/base.go (hunk 2 above)
Comment (dlarson04) [reply core/base/base.go:149]: This was done to avoid 2 go funcs (this one and communications.ResendNotifications()) both calling Register since there wasn't anyway to guarantee which one ran first. This change makes it so only the ResendNotifications() method will handle the register for the ESS

**Review by LiilyZhang — APPROVED** [review]

**Review by dabooz — APPROVED** [review]

**Comment by dlarson04** [comment]
Code changes were integrated with a different PR with some other work
